ergo-nipopow/src/nipopow_proof.rs

```diff
@@ -0,0 +1,215 @@
+    pub k: u32,
+    /// Proof prefix headers
+    #[serde(rename = "prefix")]
+    pub prefix: Vec<PoPowHeader>,
+    /// First header of the suffix
+    #[serde(rename = "suffixHead")]
+    pub suffix_head: PoPowHeader,
+    /// Tail of the proof suffix headers
+    #[serde(rename = "suffixTail")]
+    pub suffix_tail: Vec<Header>,
+}
+
+impl NipopowProof {
+    /// Create new proof instance
+    pub fn new(
+        m: u32,
+        k: u32,
+        prefix: Vec<PoPowHeader>,
+        suffix_head: PoPowHeader,
+        suffix_tail: Vec<Header>,
+    ) -> Result<NipopowProof, NipopowProofError> {
+        if k >= 1 {
+            Ok(NipopowProof {
+                popow_algos: NipopowAlgos::default(),
+                m,
+                k,
+                prefix,
+                suffix_head,
+                suffix_tail,
+            })
+        } else {
+            Err(NipopowProofError::ZeroKParameter)
+        }
+    }
+
+    /// Implementation of the ≥ algorithm from [`KMZ17`], see Algorithm 4
+    ///
+    /// [`KMZ17`]: https://fc20.ifca.ai/preproceedings/74.pdf
+    pub fn is_better_than(&self, that: &NipopowProof) -> Result<bool, NipopowProofError> {
+        if self.is_valid() && that.is_valid() {
+            if let Some(lca) = self.popow_algos.lowest_common_ancestor(
+                &self.headers_chain().collect::<Vec<_>>(),
+                &that.headers_chain().collect::<Vec<_>>(),
+            ) {
+                let self_headers = self
+                    .headers_chain()
+                    .filter(|h| h.height > lca.height)
+                    .collect::<Vec<_>>();
+                let that_headers = that
+                    .headers_chain()
+                    .filter(|h| h.height > lca.height)
+                    .collect::<Vec<_>>();
+                Ok(self.popow_algos.best_arg(&self_headers, self.m)?
+                    > self.popow_algos.best_arg(&that_headers, self.m)?)
+            } else {
+                Ok(false)
+            }
+        } else {
+            Ok(self.is_valid())
+        }
+    }
+
+    fn is_valid(&self) -> bool {
+        self.has_valid_connections() && self.has_valid_heights()
+    }
+
+    /// Checks the connections of the blocks in the proof. Adjacent blocks should be linked either
+    /// via interlink or parent block id. Returns true if all adjacent blocks are correctly
+    /// connected.
+    pub fn has_valid_connections(&self) -> bool {
+        self.prefix
+            .iter()
+            .zip(
+                self.prefix
+                    .iter()
+                    .skip(1)
+                    .chain(std::iter::once(&self.suffix_head)),
+            )
+            .all(|(prev, next)| {
+                // Note that blocks with level 0 do not appear at all within interlinks, which is
+                // why we need to check the parent block id as well.
+                next.interlinks.contains(&prev.header.id) || next.header.parent_id == prev.header.id
```

Comment:
This bit of code differs from the `ergo` reference implementation, which only contains the condition on the left arm of `||`. I found it when running this through the autolykos chain-generation method, which generates blocks with level 0. Note that the chain-generation used in `ergo` always generates blocks with positive-level.

I think the change also needs to be made in the reference implementation.

Reply:
Good call! Please, make a PR for ergo with this added check.

Reply:
Sure, will do.